target_hdfs/utils/hdfs.py

```diff
@@ -0,0 +1,122 @@
+    logger.debug(f"Uploading file from HDFS: {source_path_hdfs} ")
+    pa.fs.copy_files(
+        source_path_hdfs,
+        local_path,
+        source_filesystem=get_hdfs_client(),
+        destination_filesystem=pa.fs.LocalFileSystem(),
+    )
+    logger.debug(f"File {source_path_hdfs} downloaded from hdfs to {local_path} ")
+
+
+def upload_to_hdfs(local_file: str, destination_path_hdfs: str) -> None:
+    """Upload a local file to HDFS."""
+    logger.debug(f"Uploading file to HDFS: {destination_path_hdfs} ")
+    pa.fs.copy_files(
+        local_file,
+        destination_path_hdfs,
+        source_filesystem=pa.fs.LocalFileSystem(),
+        destination_filesystem=get_hdfs_client(),
+    )
+    logger.info(f"File {destination_path_hdfs} uploaded to HDFS")
+    replace_old_file_with_new_file(destination_path_hdfs)
+
+
+def replace_old_file_with_new_file(new_file_path: str) -> None:
+    """Replace the old file with the new file in HDFS."""
+    hdfs_client = get_hdfs_client()
+    hdfs_client.move(new_file_path, new_file_path.replace("_new", ""))
+
+
+def set_file_as_old(file_path: str) -> None:
+    """Replace the old file with the new file in HDFS."""
+    if not file_path.endswith("_old"):
+        hdfs_client = get_hdfs_client()
+        hdfs_client.move(file_path, f"{file_path}_old")
+
+
+def get_files(hdfs_path: str, extension: str = ".parquet") -> list[FileInfo]:
+    """Get all parquet files in a given HDFS path."""
+    hdfs_client = get_hdfs_client()
+    if hdfs_client.get_file_info(hdfs_path).type == FileType.NotFound:
+        return []
+    file_list = hdfs_client.get_file_info(pa.fs.FileSelector(hdfs_path, recursive=True))
+    return [file for file in file_list if file.base_name.endswith(extension)]
+
+
+def get_most_recent_file(hdfs_path: str) -> FileInfo:
+    """Get the most recent modified parquet file in a given HDFS path."""
+    files = get_files(hdfs_path, extension=".parquet_old")
+    assert (
+        len(files) <= 1
+    ), "There is more than one old file in the HDFS path which is not expected"
+    if files:
+        return files[0]
+    files = get_files(hdfs_path)
+    return max(files, key=lambda file: file.mtime) if files else None
+
+
+def read_most_recent_file(
+    hdfs_file_path: str, pyarrow_schema: pa.Schema
+) -> pa.Table | None:
+    """Read the last file from HDFS."""
+    most_recent_file = get_most_recent_file(hdfs_file_path)
+    # Force creates a new file if the last file is larger than 85% of the HDFS block size or does not exist
+    if not most_recent_file or (
+        most_recent_file.size >= get_hdfs_block_size() * 0.85
+        and most_recent_file.extension != ".parquet_old"
+    ):
+        return None
+    with NamedTemporaryFile("wb") as tmp_file:
+        download_from_hdfs(most_recent_file.path, tmp_file.name)
+        parquet_df = pa.parquet.read_table(tmp_file.name)
+        if parquet_df.schema != pyarrow_schema:
+            logger.warning(
```

Comment:
I think it's better to fail here instead of doing some new files resulting mixed schemas.